crates/typst-library/src/text/mod.rs

```diff
@@ -766,35 +790,97 @@ impl PlainText for Packed<TextElem> {
 }
 
 /// A lowercased font family like "arial".
-#[derive(Clone, Eq, PartialEq, Hash)]
-pub struct FontFamily(EcoString);
+#[derive(Debug, Clone)]
+pub struct FontFamily {
+    name: EcoString,
+    coverage: Option<Regex>,
+}
+
+impl Hash for FontFamily {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+        self.coverage_str().hash(state);
+    }
+}
+
+impl PartialEq for FontFamily {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name && self.coverage_str() == other.coverage_str()
+    }
+}
 
 impl FontFamily {
     /// Create a named font family variant.
     pub fn new(string: &str) -> Self {
-        Self(string.to_lowercase().into())
+        Self::new_with_coverage(string, None).unwrap()
+    }
+
+    pub fn new_with_coverage(
+        string: &str,
+        coverage: Option<Value>,
+    ) -> HintedStrResult<Self> {
+        let coverage = match coverage {
+            Some(Value::Str(name)) => {
+                if &*name == "latin-in-cjk" {
+                    static LATIN_IN_CJK_REGEX: LazyLock<Regex> = LazyLock::new(|| {
+                        Regex::new( "[^\u{00B7}\u{2013}\u{2014}\u{2018}\u{2019}\u{201C}\u{201D}\u{2025}-\u{2027}\u{2E3A}]").unwrap()
+                    });
+                    Some(LATIN_IN_CJK_REGEX.clone())
+                } else {
+                    bail!("unknown character set");
+                }
+            }
+            Some(coverage) => {
+                let regex = coverage.cast::<Regex>()?;
+                let ast = regex_syntax::ast::parse::Parser::new()
+                    .parse(regex.as_str())
+                    .unwrap();
+                match ast {
+                    regex_syntax::ast::Ast::ClassBracketed(..)
+                    | regex_syntax::ast::Ast::ClassUnicode(..)
+                    | regex_syntax::ast::Ast::ClassPerl(..)
+                    | regex_syntax::ast::Ast::Dot(..)
+                    | regex_syntax::ast::Ast::Literal(..) => (),
+                    _ => bail!("invalid regex in font coverage"),
```

Comment:
I don't think we need to restrict the regex in this way, to be honest. If the regex is multi-char, it just won't be a match, that's on the user. Perhaps attention should be drawn to this on the docs, but matching on the regex syntax seems a bit bug-prone.

Reply:
The `regex_syntax` crate is exactly the underlying component used by the `regex` crate. Why is it bug-prone? 

I restrict the regex syntax based on the following concerns:
- The bracket syntax in regex alone is flexible enough, users can even construct set union or intersections like `[\d&&\p{ascii}]`. Syntax like alternative or repeat do not worth the trouble in this context.
- If we want to implement sophisticated fallback algorithm like in [CSS](https://drafts.csswg.org/css-fonts/#cluster-matching) in the future, then the restricted regex syntax can be easily transformed to codepoints set without introduce API breakage.

Reply:
I meant bug-prone in the sense that we need to decide which regex syntax is useful. But I can see how it's useful to reserve the ability to transform bijectively to/from a codepoint set. However, I think we then need to document which parts of the regex syntax are permitted.